bgpd/bgp_linkstate_tlv.c

```diff
@@ -0,0 +1,1090 @@
+			case BGP_LS_TLV_IGP_ROUTER_ID_ISIS_PSEUDOWIRE_SIZE:
+			case BGP_LS_TLV_IGP_ROUTER_ID_OSPF_NON_PSEUDOWIRE_SIZE:
+			case BGP_LS_TLV_IGP_ROUTER_ID_OSPF_PSEUDOWIRE_SIZE:
+				SET_FLAG(node_descr->flags,
+					 BGP_NLRI_TLV_NODE_DESCR_IGP_ROUTER_ID);
+				memcpy(&node_descr->igp_router_id, pnt, length);
+				node_descr->igp_router_id_size = length;
+				pnt += length;
+				break;
+			default:
+				zlog_err(
+					"%s: received invalid IGP Router-ID length %hu",
+					__func__, length);
+				return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+			}
+			break;
+		default:
+			zlog_err(
+				"%s: received an unexpected node descriptor TLV %hu",
+				__func__, type);
+			return BGP_NLRI_PARSE_ERROR_PREFIX_LINKSTATE;
+		}
+	}
+
+	return BGP_NLRI_PARSE_OK;
+}
+
+static size_t bgp_linkstate_nlri_node_descriptor_display(
+	struct bgp_ls_nlri_node_descr_tlv *node_descr, char *buf, size_t size,
+	bool multiline)
+{
+	bool first = true;
+
+	if (CHECK_FLAG(node_descr->flags, BGP_NLRI_TLV_NODE_DESCR_LOCAL_NODE))
+		snprintf(buf, size, "%sLocal%s", multiline ? "\n" : " ",
+			 multiline ? "  " : "{");
+	else if (CHECK_FLAG(node_descr->flags,
+			    BGP_NLRI_TLV_NODE_DESCR_REMOTE_NODE))
+		snprintf(buf, size, "%sRemote%s", multiline ? "\n" : " ",
+			 multiline ? " " : "{");
+	size -= strlen(buf);
+	buf += strlen(buf);
+
+	if (CHECK_FLAG(node_descr->flags,
+		       BGP_NLRI_TLV_NODE_DESCR_AUTONOMOUS_SYSTEM)) {
+		snprintf(buf, size, "%sAS:%u", first ? "" : " ",
+			 node_descr->autonomous_system);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(node_descr->flags, BGP_NLRI_TLV_NODE_DESCR_BGP_LS_ID)) {
+		snprintf(buf, size, "%sID:%u", first ? "" : " ",
+			 node_descr->bgp_ls_id);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(node_descr->flags, BGP_NLRI_TLV_NODE_DESCR_AREA_ID)) {
+		snprintf(buf, size, "%sArea:%u", first ? "" : " ",
+			 node_descr->area_id);
+		size -= strlen(buf);
+		buf += strlen(buf);
+		first = false;
+	}
+	if (CHECK_FLAG(node_descr->flags,
+		       BGP_NLRI_TLV_NODE_DESCR_IGP_ROUTER_ID)) {
+		switch (node_descr->igp_router_id_size) {
+		case BGP_LS_TLV_IGP_ROUTER_ID_ISIS_NON_PSEUDOWIRE_SIZE:
+			snprintfrr(buf, size, "%sRtr:%pSY", first ? "" : " ",
+				   (uint8_t *)&node_descr->igp_router_id);
+			break;
+		case BGP_LS_TLV_IGP_ROUTER_ID_ISIS_PSEUDOWIRE_SIZE:
+			snprintfrr(buf, size, "%sRtr:%pPN", first ? "" : " ",
+				   (uint8_t *)&node_descr->igp_router_id);
+			break;
+		case BGP_LS_TLV_IGP_ROUTER_ID_OSPF_NON_PSEUDOWIRE_SIZE:
+			snprintfrr(buf, size, "%sRtr:%pI4", first ? "" : " ",
+				   (in_addr_t *)&node_descr->igp_router_id);
+			break;
+		case BGP_LS_TLV_IGP_ROUTER_ID_OSPF_PSEUDOWIRE_SIZE:
+			snprintfrr(
+				buf, size, "%sRtr:%pI4:%pI4", first ? "" : " ",
+				(in_addr_t *)&node_descr->igp_router_id,
+				((in_addr_t *)&node_descr->igp_router_id + 1));
+			break;
+		}
+
+		size -= strlen(buf);
+		buf += strlen(buf);
+	}
+
```

Comment:
the list of unknown tlvs should be kept and displayed: {type %u, len %u}

Reply:
see https://github.com/FRRouting/frr/pull/12649#discussion_r1169925019